README
-- GrSL - the GSL random Sampling Library --

GrSL is an extension of the GNU Scientific Library (GSL) that aims to
provide efficient algorithms for sequential random sampling.  It is
built with the aim of eventually being incorporated into GSL  proper,
but is not yet an official part of the GSL.

Its API is designed for close compatibility with GSL, with the aim of
minimising (or hopefully not requiring) any changes to users' code if/
when it is incorporated into the main GSL library.

The project is currently in an extremely preliminary stage.
Contributions of advice, code and feedback are all welcome.

GrSL is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License, version 3 or (at your
option) any later version.

--REFERENCES--

  Nair KA (1990) 'An improved algorithm for ordered sequential
    random sampling.'  ACM T. Math. Softw. 16(3): 269--274.

  Vitter JS (1984) 'Faster methods for random sampling.'
    Commun. ACM 27(7): 703--718

  Vitter JS (1987) 'An efficient algorithm for sequential random
    sampling.'  ACM T. Math. Softw. 13(1): 58--67.


NEWS
** 2010/05/11 **

  GrSL only started development today, and is therefore so new that there is
  hardly any news about it. :-)

  The current very limited functionality implements Algorithm A from Vitter
  (1984, 1987) and has a couple of small demos of it at work.


grsl-test.c
/* grsl-test.c
 *
 * ---------------------------------------------------------------------
 * A small and very daft little set of tests for GrSL's sampling
 * functions.
 * ---------------------------------------------------------------------
 *
 * Copyright (C) 2010 Joseph Rushton Wakeling
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <config.h>
#include <gsl/gsl_errno.h>
#include <gsl/gsl_randist.h>
#include <gsl/gsl_sampling.h>

void grsl_test_simple(const gsl_sampler *s, const gsl_rng *r, size_t n, size_t N)
{
  size_t i, current_record, selected_record;

  gsl_sampler_init(s,r,n,N);
  current_record = 1;

  printf("%s, %zu from %zu:\n",s->algorithm->name, n, N);

  for ( i = 0 ; i < s->sample->total ; ++i)
    {
      selected_record = gsl_sampler_select(s,r,&current_record);
      printf("\tselected record %zu.",selected_record);
      printf("\trecords remaining: %zu.",s->records->remaining);
      printf("\tremaining to select: %zu.\n",s->sample->remaining);
    }


}

void grsl_test_aggregate(const gsl_sampler *s, const gsl_rng *r, size_t n, size_t N,
                         size_t repeats)
{
  size_t i, j, current_record, selected_record;
  size_t *record_count;
  clock_t start_time, end_time;

  record_count = malloc(N*sizeof(size_t));

  for(i=0;i<N;++i)
    record_count[i] = 0;

  printf("%s, %zu from %zu x %zu:\n",s->algorithm->name, n, N, repeats);

  start_time = clock();

  for(i=0;i<repeats;++i)
    {
      gsl_sampler_init(s,r,n,N);
      current_record = 0;  /* note that we start with record 0 this time,
                              as we are going to be picking from an array ... */

      for ( j = 0 ; j < s->sample->total ; ++j)
        {
          selected_record = gsl_sampler_select(s,r,&current_record);
          record_count[selected_record]++;
        }
    }

  end_time = clock();

  for(i=0;i<N;++i)
    printf("\trecord %zu was picked %zu times.\n",i+1,record_count[i]);

  printf("\t\tsampling completed in %g seconds.\n",
         ((double) (end_time-start_time))/CLOCKS_PER_SEC);

  free(record_count);
}

int main(int argc, char *argv[])
{
  size_t i;
  gsl_sampler *s = gsl_sampler_alloc(gsl_sampler_vitter_a);
  gsl_sampler *sd = gsl_sampler_alloc(gsl_sampler_vitter_d);
  gsl_rng *r = gsl_rng_alloc(gsl_rng_mt19937);
  double *dest, *src;
  time_t ranseed;
  clock_t start_time, end_time;

  if(argc>1)
    ranseed = atoi(argv[1]);
  else
    time(&ranseed);

  gsl_rng_set(r, ranseed);

  printf("[Random seed set to %lu.]\n\n",(unsigned long int) ranseed);

  printf("Hello! I'm GrSL, the GSL random Sampling Library.\n\n");

  printf("Basically, I'm a cheeky little scamp from out of town with dreams of\n");
  printf("joining the GNU Scientific Library when I grow up.\n\n");

  printf("Let me show you what I can do so far.\n\n");

  printf("First, I'm going to make some samples of 5 records out of 100.\n\n");

  grsl_test_simple(s, r, 5, 100);
  grsl_test_simple(sd, r, 5, 100);

  printf("\n");
  printf("Now, I'm going to make a sample of 5 records out of 100, but do so\n");
  printf("10 million times.  Then you can count how many times each record gets\n");
  printf("picked.  (This is just a stupid way of checking for obvious bias:-)\n\n");

  grsl_test_aggregate(s, r, 5, 100, 10000000);
  grsl_test_aggregate(sd, r, 5, 100, 10000000);

  printf("\n");
  printf("Next up, we provide a comparison of the gsl_ran_choose function with\n");
  printf("the new gsl_sampler_choose, which provides the same functionality but\n");
  printf("employing the new sampling algorithms provided by GrSL.\n\n");

  printf("We are going to pick 100,000 records from an array of 10 million.\n");

  src = malloc(10000000*sizeof(*src));
  dest = malloc(100000*sizeof(*dest));

  for(i=0;i<10000000;++i)
    src[i] = i+1;

  printf("\tgsl_ran_choose:\n");

  start_time = clock();
  gsl_ran_choose(r, dest, 100000, src, 10000000, sizeof(double));
  end_time = clock();

  printf("\t\tfinished in %g seconds.\n",
         ((double) (end_time-start_time))/CLOCKS_PER_SEC);

  printf("\tgsl_sampler_choose:\n");

  start_time = clock();
  gsl_sampler_choose(s, r, dest, 100000, src, 10000000, sizeof(double));
  end_time=clock();

  printf("\t\tfinished in %g seconds with %s.\n",
         ((double) (end_time-start_time))/CLOCKS_PER_SEC, s->algorithm->name);

  start_time = clock();
  gsl_sampler_choose(sd, r, dest, 100000, src, 10000000, sizeof(double));
  end_time=clock();

  printf("\t\tfinished in %g seconds with %s.\n",
         ((double) (end_time-start_time))/CLOCKS_PER_SEC, sd->algorithm->name);

  free(dest);
  free(src);

  gsl_sampler_free(s);
  gsl_sampler_free(sd);
  gsl_rng_free(r);

  return EXIT_SUCCESS;
}


sampling/sampling.c
/* sampling/sampling.c
 *
 * Copyright (C) 2010 Joseph Rushton Wakeling
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>
#include <gsl/gsl_errno.h>
#include <gsl/gsl_rng.h>
#include <gsl/gsl_sampling.h>

gsl_sampler *
gsl_sampler_alloc(const gsl_sampling_algorithm *A)
{
  gsl_sampler *s = malloc(sizeof(gsl_sampler));

  if (s == 0)
    {
      GSL_ERROR_VAL ("failed to allocate space for sampler struct",
                     GSL_ENOMEM, 0);
    };

  s->state = malloc(A->size);

  if (s->state == 0)
    {
      free (s);         /* exception in constructor, avoid memory leak */

      GSL_ERROR_VAL ("failed to allocate space for sampler state",
                     GSL_ENOMEM, 0);
    };

  s->records = malloc(sizeof(gsl_sampling_records));

  if (s->records == 0)
    {
      free(s->state);
      free(s);

      GSL_ERROR_VAL ("failed to allocate space for records",
                     GSL_ENOMEM, 0);
    }

  s->sample = malloc(sizeof(gsl_sampling_records));

  if (s->sample == 0)
    {
      free(s->records);
      free(s->state);
      free(s);
      GSL_ERROR_VAL ("failed to allocate space for sample",
                     GSL_ENOMEM, 0);
    }

  s->algorithm = A;

  s->sample->remaining = s->sample->total = s->records->remaining = s->records->total = 0;

  return s;
}

void
gsl_sampler_free(gsl_sampler * s)
{
  RETURN_IF_NULL(s);
  free(s->sample);
  free(s->records);
  free(s->state);
  free(s);
}

int
gsl_sampler_init(const gsl_sampler * s, const gsl_rng *r, size_t sample_size,
                 size_t records)
{
  if ( sample_size > records )
    {
      GSL_ERROR ("Sample size cannot be greater than the number of records!",
                 GSL_EINVAL) ;
    }

  s->sample->remaining = s->sample->total = sample_size;
  s->records->remaining = s->records->total = records;
  (s->algorithm->init) (s->state, s->sample, s->records, r);

  return GSL_SUCCESS;
}

/* Include a copy of ... well, copy ... for gsl_sampler_choose to use.
   TODO: check if you can actually use the one in gsl's randist/shuffle.c.
 */
static inline void
copy (void * dest, size_t i, void * src, size_t j, size_t size)
{
  register char * a = size * i + (char *) dest ;
  register char * b = size * j + (char *) src ;
  register size_t s = size ;

  do
    {
      *a++ = *b++;
    }
  while (--s > 0);
}

/* Reimplements the gsl_ran_choose function in randist/shuffle.c of the
   GSL to use GrSL sampling algorithms to select the chosen subset.

   Assuming that sample size k is << the number n of records, this runs
   in approximately 2/3 the time of gsl_ran_choose when using the simple
   Algorithm A.  (Both algorithms are O(n) in running time but Algorithm
   A requires only k random variates to be generated.)

   Run with the more efficient Algorithm D, which runs in o(k) time and
   requires only about k random variates, the savings can be considerable.
 */
int
gsl_sampler_choose(const gsl_sampler * s, const gsl_rng * r, void * dest,
                   size_t k, void * src, size_t n, size_t size)
{
  size_t i, current_record = 0, selected_record;

  if ( k > n )
    {
      GSL_ERROR ("k is greater than n, cannot sample more than n items",
                 GSL_EINVAL) ;
    }

  gsl_sampler_init(s, r, k, n);

  for(i=0;i<k;++i)
    {
      selected_record = gsl_sampler_select(s, r, &current_record);
      copy(dest, i, src, selected_record, size);
    }

  return GSL_SUCCESS;
}


sampling/vitter_a.c
/* sampling/vitter_a.c
 *
 * ---------------------------------------------------------------------
 * Provides an implementation of 'Algorithm A' described in
 *
 *   Vitter JS (1984) 'Faster methods for random sampling'.
 *     Commun. ACM 27(7): 703--718
 *
 *   Vitter JS (1987) 'An efficient algorithm for sequential random
 *     sampling.'  ACM T. Math. Softw. 13(1): 58--67.
 * ---------------------------------------------------------------------
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <gsl/gsl_rng.h>
#include <gsl/gsl_sampling.h>

static size_t
vitter_a_skip(const gsl_rng * r, size_t * const remaining_records,
              size_t * const remaining_samples)
{
  size_t S;
  double top, V, quot;

  if (*remaining_samples == 1)
    {
      S = gsl_rng_uniform_int(r, *remaining_records);
    }
  else
    {
      S = 0;
      top = *remaining_records - *remaining_samples;
      V = gsl_rng_uniform_pos(r);
      quot = top/(*remaining_records);

      while (quot > V)
        {
          ++S;
          --top;
          --(*remaining_records);
          quot *= top/(*remaining_records);
        }
    }

  --(*remaining_records);
  --(*remaining_samples);

  return S;
}

static const gsl_sampler vitter_a = {"vitter_a", &vitter_a_skip};

const gsl_sampler *gsl_sampler_vitter_a = &vitter_a;


sampling/gsl_sampling.h
/* sampling/gsl_sampling.h
 *
 * ---------------------------------------------------------------------
 * Header file for GrSL, designed to facilitate long-term incorporation
 * into the GNU Scientific Library.
 * ---------------------------------------------------------------------
 *
 * Copyright (C) 2010 Joseph Rushton Wakeling
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef __GSL_SAMPLING_H__
#define __GSL_SAMPLING_H__
#include <gsl/gsl_types.h>
#include <gsl/gsl_errno.h>
#include <gsl/gsl_inline.h>
#include <gsl/gsl_rng.h>

#undef __BEGIN_DECLS
#undef __END_DECLS
#ifdef __cplusplus
# define __BEGIN_DECLS extern "C" {
# define __END_DECLS }
#else
# define __BEGIN_DECLS /* empty */
# define __END_DECLS /* empty */
#endif

__BEGIN_DECLS

typedef struct
  {
    size_t total;
    size_t remaining;
  }
gsl_sampling_records;

typedef struct
  {
    const char *name;
    size_t size;
    void (*init) (void * vstate, const gsl_sampling_records * const sample,
                  const gsl_sampling_records * const records, const gsl_rng *r);
    size_t (*skip) (void * vstate, gsl_sampling_records * const sample,
                    gsl_sampling_records * const records, const gsl_rng *r);
  }
gsl_sampling_algorithm;

typedef struct
  {
    const gsl_sampling_algorithm *algorithm;
    gsl_sampling_records *sample;
    gsl_sampling_records *records;
    void *state;
  }
gsl_sampler;


GSL_VAR const gsl_sampling_algorithm *gsl_sampler_vitter_a;
GSL_VAR const gsl_sampling_algorithm *gsl_sampler_vitter_d;
GSL_VAR const gsl_sampling_algorithm *gsl_sampler_nair_e;


gsl_sampler *
gsl_sampler_alloc(const gsl_sampling_algorithm *A);

void
gsl_sampler_free(gsl_sampler * s);

int
gsl_sampler_init(const gsl_sampler *s, const gsl_rng *r, size_t sample_size,
                 size_t records);

int
gsl_sampler_choose(const gsl_sampler * s, const gsl_rng * r, void * dest,
                   size_t k, void * src, size_t n, size_t size);


#ifdef HAVE_INLINE

INLINE_FUN size_t
gsl_sampler_skip (const gsl_sampler * s, const gsl_rng * r)
{
  size_t S;
  if (s->records->remaining == 0)
    {
      GSL_ERROR_VAL ("No more records left to sample.",
                     GSL_EINVAL, 0) ;
    }
  else if (s->sample->remaining == 0)
    {
      GSL_ERROR_VAL ("Sample already contains the required number of records.",
                     GSL_EINVAL, 0) ;
    }

  S = (s->algorithm->skip) (s->state, s->sample, s->records, r);

  --(s->sample->remaining);
  s->records->remaining -= (S+1);

  return S;
}

INLINE_FUN size_t
gsl_sampler_select(const gsl_sampler * s, const gsl_rng * r,
                   size_t * const current_record)
{
  *current_record += gsl_sampler_skip(s, r);
  return (*current_record)++;
}

#endif /* HAVE_INLINE */

__END_DECLS

#endif /* __GSL_SAMPLING_H__ */


sampling/vitter.c
/* sampling/vitter.c
 *
 * ---------------------------------------------------------------------
 * Provides an implementation of Algorithms A and D introduced by
 * Jeffrey Scott Vitter in the following articles:
 *
 *   Vitter JS (1984) 'Faster methods for random sampling'.
 *     Commun. ACM 27(7): 703--718
 *
 *   Vitter JS (1987) 'An efficient algorithm for sequential random
 *     sampling.'  ACM T. Math. Softw. 13(1): 58--67.
 * ---------------------------------------------------------------------
 *
 * Copyright (C) 2010 Joseph Rushton Wakeling
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <math.h>
#include <stdbool.h>
#include <gsl/gsl_randist.h>
#include <gsl/gsl_rng.h>
#include <gsl/gsl_sampling.h>

/* Vitter (1984) introduces Algorithm A as a component part of the
   still-more-efficient Algorithm D.

   By itself it runs in O(N) time, where N is the total number of
   records, but generates only n random variates, where n is the
   number of records to be sampled.

   Algorithm D uses this method when the number of remaining samples
   to be taken is greater than a certain proportion (0.05--0.15) of
   the total number of remaining records to be sampled from.
 */
void
vitter_a_init(void * vstate, const gsl_sampling_records * const sample,
              const gsl_sampling_records * const records, const gsl_rng *r)
{
  /* Algorithm A does not require any initialisation :-)
     For the same reason, no vitter_a_state_t is defined.
   */
}

/*
   A few notes on the implementation of the Algorithm A skip function:

     * Use of gsl_rng_uniform_pos to set value of V.  Vitter (1984,
       1987) assumes random variates are in the open interval (0,1).
       The requirement from this seems to stem (in Algorithm A at
       least) from the fact (AFAICS, do not trust me on this one:-)
       that when top == 0 (and hence quot==0), it is _essential_ to
       pick the next record, i.e. we cannot allow the rare but
       possible case that V==0.

         -- Note that this could allow a saving in the number of
            required random variates by including an if(top>0)
            statement.  However, the cost of the continually-being-
            checked if() statement turns out to be greater than the
            savings from the (apparently rare) cases when top==0.

     * Use of gsl_rng_uniform_int if the number of remaining samples
       is only 1.  Vitter (1984) simply calls for the truncation (i.e.
       integer part) of the product of a random variate in (0,1) and
       the number N of remaining records.  Vitter (1987) gives a more
       complicated expression,

           S := TRUNC( ROUND(Nreal) * UNIFORMRV() )

       (where Nreal is a real-valued variable equal to the integer N),
       in order to avoid the out-of-range possibility of S being
       exactly equal to N.  All these appear to merely be complicated
       ways of describing generating a uniformly-distributed integer
       in the range [0, N-1], which is what gsl_rng_uniform_int
       provides.

     * Alteration of s->records->remaining and s->samples->remaining:
       see if this can be removed entirely outside these individual
       skip functions and placed entirely in gsl_sampler_skip.
 */
static size_t
vitter_a_skip(void * vstate, gsl_sampling_records * const sample,
              gsl_sampling_records * const records, const gsl_rng *r)
{
  register size_t S;
  register double V, quot, top;

  if (sample->remaining == 1)
    {
      S = gsl_rng_uniform_int(r, records->remaining);
    }
  else
    {
      S = 0;
      top = records->remaining - sample->remaining;
      quot = top/(records->remaining);
      V = gsl_rng_uniform_pos(r);

      while (quot > V)
        {
          ++S;
          quot *= (top - S) / (records->remaining - S);
        }
    }

  return S;
}

static const gsl_sampling_algorithm vitter_a =
{"vitter_a",                  /* name */
 0,                           /* size */
 &vitter_a_init,              /* init */
 &vitter_a_skip               /* skip */
};

const gsl_sampling_algorithm *gsl_sampler_vitter_a = &vitter_a;


/* Algorithm D, introduced in Vitter (1984), requires only ~n random
   variates to be generated and runs in O(n) time.  This implementation
   follows the more efficient version introduced in Vitter (1987).

   The method calls Algorithm A to generate the skip size when the
   number of remaining samples to be taken is greater than a certain
   proportion alpha of the total number of remaining records.  This
   implementation follows Vitter (1987) in taking alpha = 1/13.

   Brief and entirely inadequate testing on the present author's part
   suggests that this is indeed an optimal choice. :-)

   The algorithm was further refined by Nair (1990) whose Algorithm E
   takes advantage of some cases where skip values of Algorithm A can
   be calculated exactly in a single step.
 */
typedef struct
  {
    double Vprime;
    bool use_algorithm_a;
  }
vitter_d_state_t;

/* As per Vitter (1984, 1987) we do not store the floating-point value
   of alpha but store an integer value that is 1/alpha.  Following
   Vitter's suggestion this is set to 13. */
static const short int vitter_d_alpha_inverse = 13;

/* Algorithm D stores two pieces of information: the random variate
   Vprime, which must be preserved between calls to the skip function,
   and a boolean to indicate whether or not to generate remaining
   skip values using Algorithm A. */
void
vitter_d_init(void * vstate, const gsl_sampling_records * const sample,
              const gsl_sampling_records * const records, const gsl_rng *r)
{
  vitter_d_state_t *state = vstate;

  /* We can save ourselves one random variate by checking at the very
     beginning whether or not the sample size is larger than the threshold
     to start using Algorithm A. */
  if ( (vitter_d_alpha_inverse * sample->remaining) > records->remaining )
    {
      state->use_algorithm_a = true;
    }
  else
    {
      state->Vprime = pow ( gsl_rng_uniform_pos(r), 1.0/(sample->remaining) ) ;
      state->use_algorithm_a = false;
    }
}

/* Algorithm D's skip function employs some clever tricks to minimise
   the number of random variates that must be generated -- if we are
   lucky, the algorithm exits with a condition such that the next
   variate in the sequence can be calculated cheaply and numerically
   directly from the one before.

   The Pascal-style description of the algorithm in Vitter (1984, 1987)
   employs multiple statements of the form exp(log(x) * y), presumably
   because Pascal lacks a built-in power function.  These have been
   transcribed into the form pow(x, y) in the present implementation.
   Thanks to Zoltan Kuscsik for suggesting out this improvement. :-)
 */
static size_t
vitter_d_skip(void * vstate, gsl_sampling_records * const sample,
              gsl_sampling_records * const records, const gsl_rng *r)
{
  size_t S;
  vitter_d_state_t *state = vstate;
  size_t top, t, limit;
  size_t qu1 = 1 + records->remaining - sample->remaining;
  double X, y1, y2, bottom;

  /* If the remainining number of sample points needed is greater than
     a certain proportion of the remaining records, we finish off using
     Algorithm A... */
  if ( (vitter_d_alpha_inverse * sample->remaining) > records->remaining )
    state->use_algorithm_a = true;

  /* ... like this. :-) */
  if ( state->use_algorithm_a )
    {
      return vitter_a_skip(NULL, sample, records, r);
    }
  else if ( sample->remaining > 1)
    {
      while ( 1 )
        {
          /* Step D2: set X and U */
          while ( 1 )
            {
              X = records->remaining * (1 - state->Vprime);
              S = trunc(X);
              if ( S < qu1 )
                break;
              else
                state->Vprime = pow ( gsl_rng_uniform_pos(r), 1.0/(sample->remaining) );
            }

          y1 = pow ( (gsl_rng_uniform_pos(r) * ((double) records->remaining)/qu1),
                     (1.0/(sample->remaining - 1)) );

          state->Vprime
            = y1 * ((-X/records->remaining)+1.0) * ( qu1/( ((double) qu1) - S ) );

          /* Step D3: if state->Vprime <= 1.0 our work is done, otherwise ... */
          if ( state->Vprime > 1.0 )
            {
              y2 = 1.0;
              top = records->remaining - 1;

              if ( sample->remaining > (S+1) )
                {
                  bottom = records->remaining - sample->remaining;
                  limit = records->remaining - S;
                }
              else
                {
                  bottom = records->remaining - (S+1);
                  limit = qu1;
                }

              for ( t = (records->remaining - 1); t >= limit; --t)
                {
                  y2 = (y2 * top)/bottom;
                  --top;
                  --bottom;
                }

              /* Step D4: decide whether or not to go right back to the start
                 of this damn while() loop ... :-) */

              if( (records->remaining/(records->remaining - X))
                    < ( y1 * pow(y2, 1.0/(sample->remaining - 1)) ) )
                {
                  /* If we're unlucky, we just have to generate a new Vprime
                     and go right back to the beginning.
                     printf("D4 fail.  "); fflush(stdout); */
                  state->Vprime
                    = pow ( gsl_rng_uniform_pos(r),  1.0/(sample->remaining) ) ;
                }
              else
                {
                  /* If we're lucky, we accept S and generate a new Vprime ...
                     printf("D4 exit: %zu\n",S); fflush(stdout); */
                  state->Vprime
                    = pow ( gsl_rng_uniform_pos(r), 1.0/(sample->remaining - 1) ) ;
                  return S;
                }
            }
          else
            {
              /* printf("D3 exit: %zu\n",S); fflush(stdout); */
              return S;
            }
        }
    }
  else
    {
      /* If only one sample point remains to be taken ... */
      return trunc ( records->remaining * state->Vprime );
    }
}

static const gsl_sampling_algorithm vitter_d =
{"vitter_d",                  /* name */
 sizeof(vitter_d_state_t),    /* size */
 &vitter_d_init,              /* init */
 &vitter_d_skip               /* skip */
};

const gsl_sampling_algorithm *gsl_sampler_vitter_d = &vitter_d;
